Threescale 7668

Patryk-Stefanski opened the pull request:
# What 
[THREESCALE-7668](https://issues.redhat.com/browse/THREESCALE-7668)
Addition of new application CR, that supports :
- Create
- Update (Name, application plan, description)
- Delete
- Suspend 

# Verification 
1. ~As this [PR](https://github.com/3scale/3scale-porta-go-client/pull/48) with new functionality hasnt been merged yet, check out [this branch](https://github.com/Patryk-Stefanski/3scale-porta-go-client/tree/THREEESCALE-7668) of porta client locally.~

2. ~Check out this 3scale-operator branch locally and add the following line of code to go.mod. (This will use the local version of the porta client setup in step1)~
```bash
replace github.com/3scale/3scale-porta-go-client v0.6.0 =>  <LOCATION_OF_LOCAL_PORTA_CLIENT_REPO>
```

4. Run the following commands to install the 3scale operator
```bash
go mod vendor 
go mod tidy 
make install
export NAMESPACE=3scale-test
oc new-project $NAMESPACE
make download
make run
```

5. For the operator to create routes we need to mock the s3 credentials secret, run the following command.
```bash
kubectl apply -f - <<EOF
---
kind: Secret
apiVersion: v1
metadata: 
  name: s3-credentials
  namespace: 3scale-test
data: 
  AWS_ACCESS_KEY_ID: UkVQTEFDRV9NRQ==
  AWS_BUCKET: UkVQTEFDRV9NRQ==
  AWS_REGION: UkVQTEFDRV9NRQ==
  AWS_SECRET_ACCESS_KEY: UkVQTEFDRV9NRQ==
type: Opaque
EOF
```

6. Deploy the API manager CR  and specify your cluster wildcardDomain.
```bash
kubectl apply -f - <<EOF
---
apiVersion: apps.3scale.net/v1alpha1
kind: APIManager
metadata: 
  name: apimanager-sample
  namespace: 3scale-test
spec: 
  system: 
    fileStorage: 
      simpleStorageService: 
        configurationSecretRef: 
          name: s3-credentials
  wildcardDomain: apps.pstefans.****.s1.devshift.org
EOF
```

7. Create the required secrets for Accounts, products and backends. You can find the admin token in system-seed secret and the admin URL  under routes in the 3scale-test namespace.
```bash
kubectl apply -f - <<EOF
---
apiVersion: v1
kind: Secret
metadata:
  name: mytenant
type: Opaque
stringData:
  adminURL: <ADMIN_URL>
  token: "<ADMIN_ACCESS_TOKEN>"
EOF

kubectl apply -f - <<EOF
---
apiVersion: v1
kind: Secret
metadata:
  name: myusername01
stringData:
  password: "123456"
EOF
```

8. Create the developerAccount, DeveloperUser, product and backend required for the application CR.
```bash 
kubectl apply -f - <<EOF
---
apiVersion: capabilities.3scale.net/v1beta1
kind: DeveloperUser
metadata:
  name: developeruser01
  namespace: 3scale-test
spec:
  developerAccountRef:
    name: developeraccount01
  email: myusername01@example.com
  passwordCredentialsRef:
    name: myusername01
  providerAccountRef:
    name: mytenant
  role: admin
  username: myusername01
EOF

kubectl apply -f - <<EOF
---
apiVersion: capabilities.3scale.net/v1beta1
kind: DeveloperAccount
metadata:
  name: developeraccount01
  namespace: 3scale-test
spec:
  orgName: pstefans3
  providerAccountRef:
    name: mytenant
EOF


kubectl apply -f - <<EOF
---
apiVersion: capabilities.3scale.net/v1beta1
kind: Backend
metadata:
  name: backend1-cr
  namespace: 3scale-test
spec:
  mappingRules:
    - httpMethod: GET
      increment: 1
      last: true
      metricMethodRef: hits
      pattern: /
  name: backend1
  privateBaseURL: 'https://api.example.com'
  systemName: backend1
EOF

kubectl apply -f - <<EOF
---
apiVersion: capabilities.3scale.net/v1beta1
kind: Product
metadata:
  name: product1-cr
  namespace: 3scale-test
spec:
  applicationPlans:
    plan01:
      name: "My Plan 01"
      limits:
        - period: month
          value: 300
          metricMethodRef:
            systemName: hits
            backend: backend1
    plan02:
      name: "My Plan 02"
      limits:
        - period: month
          value: 300
          metricMethodRef:
            systemName: hits
            backend: backend1
  name: product1
  backendUsages:
    backend1:
      path: /
EOF
```

9.Create an Invalid Application CR , status should give out about invalid account and product CR names.
```bash
kubectl apply -f - <<EOF
---
apiVersion: capabilities.3scale.net/v1beta1
kind: Application
metadata:
  name: example
  namespace: 3scale-test
spec:
  accountCR: 
    name: developeraccount02
  applicationPlanName: plan03
  productCR: 
    name: product2-cr
  name: testApp
  description: further testing
EOF
```

10.  Update Application Resource , it should give out about the plan being used not belonging to the product.
```bash
kubectl apply -f - <<EOF
---
apiVersion: capabilities.3scale.net/v1beta1
kind: Application
metadata:
  name: example
  namespace: 3scale-test
spec:
  accountCR: 
    name: developeraccount01
  applicationPlanName: plan03
  productCR: 
    name: product1-cr
  name: testApp
  description: further testing
EOF
```

11. Update Application resource once again, this time the status should report ready: true with no error message. Check in 3scale UI that app has been created.
```bash
kubectl apply -f - <<EOF
---
apiVersion: capabilities.3scale.net/v1beta1
kind: Application
metadata:
  name: example
  namespace: 3scale-test
spec:
  accountCR: 
    name: developeraccount01
  applicationPlanName: plan01
  productCR: 
    name: product1-cr
  name: testApp
  description: further testing
EOF
```

11. Update the application one last time to check that updating the name , description , plan and suspending works.Check that the application update is reflected in 3scale UI and that the app has gone into the suspended state.
```bash
kubectl apply -f - <<EOF
---
apiVersion: capabilities.3scale.net/v1beta1
kind: Application
metadata:
  name: example
  namespace: 3scale-test
spec:
  accountCR: 
    name: developeraccount01
  applicationPlanName: plan02
  productCR: 
    name: product1-cr
  name: testApp12 
  description: further testing12
  suspend: true
EOF
```

eguzki commented on apis/capabilities/v1beta1/application_types.go:
```diff
@@ -0,0 +1,140 @@
+
+// ApplicationSpec defines the desired state of Application
+type ApplicationSpec struct {
+	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
```
missing provider account 

```
    // ProviderAccountRef references account provider credentials                        
    // +optional                                                                         
    ProviderAccountRef *corev1.LocalObjectReference `json:"providerAccountRef,omitempty"`

```

Patryk-Stefanski replied:
Yeah, the application uses the developer account for its reference. Not sure where I should document that at the moment, could you point me in the right direction?

I've added a tiny bit of logic to the externalRefCheck() that will check if both resources have the same providerAccountHost if not then it should return an error.

austincunningham replied:
Guessing this should be documented in the `doc` directory under `application-reference.md` I will make a start on it.

austincunningham replied:
added documentation
https://github.com/3scale/3scale-operator/pull/778/files#diff-539a57d01480bffad575ce52c6aac9277d2495037aee29b0309811d92e0f6855

austincunningham replied:
With the addition of the docs and the checkExternalResources() check https://github.com/3scale/3scale-operator/pull/778/commits/4711eb55d6d7be3e7d39f51d6f7867e2fa3600d3#diff-067ba2ea2f14e8822e886fc126454337bc5e488fe598f1a6f2f6ad536c572dabR259 think this is resolved.

Patryk-Stefanski commented on apis/capabilities/v1beta1/application_types.go:
```diff
@@ -0,0 +1,140 @@
+	// Important: Run "make" to regenerate code after modifying this file
+
+	//AccountCRName name of account custom resource under which the application will be created
+	AccountCRName string `json:"accountCRName"`
```
addressed

Patryk-Stefanski commented on config/samples/capabilities_v1beta1_application.yaml:
```diff
@@ -0,0 +1,7 @@
+  name: application-sample
+spec:
+  # Add fields here
+  foo: bar
```
addressed

Patryk-Stefanski commented on apis/capabilities/v1beta1/application_types.go:
```diff
@@ -0,0 +1,140 @@
+	AccountCRName string `json:"accountCRName"`
+
+	//ProductCRName of product custom resource from which the application plan will be used
+	ProductCRName string `json:"productCRName"`
```
addressed

Patryk-Stefanski commented on pkg/controller/helper/application_entity.go:
```diff
@@ -0,0 +1,50 @@
+	logger logr.Logger
+}
+
+func NewApplicationEntity(ApplicationObj *threescaleapi.Application, client *threescaleapi.ThreeScaleClient, logger logr.Logger) *ApplicationEntity {
```
Not at the minute. However it will be needed for tests later on.

austincunningham replied:
adding back used in the status reconciler

Patryk-Stefanski commented on controllers/capabilities/application_controller.go:
```diff
@@ -0,0 +1,273 @@
+	}
+
+	// get Account
+	Account := &capabilitiesv1beta1.DeveloperAccount{}
```
addressed

Patryk-Stefanski commented on controllers/capabilities/application_controller.go:
```diff
@@ -0,0 +1,273 @@
+	accountFldPath := specFldPath.Child("accountCRName")
+	productFldPath := specFldPath.Child("productCRName")
+
+	data, err := json.Marshal(accountResource)
```
Marshalling is used to check if a resource has a valid ID, if you try to check the resource ID without marshalling you will hit a nil pointer if the ID doesnt exist.

eguzki replied:
Then, check if the ID exists. Furthermore, you should check that referenced developerAccount CR and Product CR are both in valid state. If not, the Application will also be in not-ready state.

austincunningham replied:
checking for nil on ID for product and account, also checking invalid status now

Patryk-Stefanski commented:
@austincunningham would you mind addressing the remaining comments? Not sure I can do much without a cluster.

FYI some tests are failing cause we are using porta-client v0.6 which doesnt have the new changes,  think we might need a new tag for porta @eguzki

austincunningham commented on apis/capabilities/v1beta1/application_types.go:
```diff
@@ -0,0 +1,141 @@
+	// Important: Run "make" to regenerate code after modifying this file
+
+	//AccountCRName name of account custom resource under which the application will be created
+	AccountCRName *corev1.LocalObjectReference `json:"accountCRName"`
```
@Patryk-Stefanski 
Just starting to look at this today
Just wondering if the application CR's in the verification steps are correct
```
kubectl apply -f - <<EOF
---
apiVersion: capabilities.3scale.net/v1beta1
kind: Application
metadata:
  name: example
  namespace: 3scale-test
spec:
  accountCRName: developeraccount01
  applicationPlanName: plan02
  productCRName: product1-cr
  name: testApp12 
  description: further testing12
  suspend: true
EOF ```
because you are using LocalObjectReference as a type
accountCRName, productCRName would be something like
```bash
  accountCRName: 
    name: test
  productCRName:
    name: test
```

austincunningham replied:
Yea it looks like verification steps need updating
![image](https://user-images.githubusercontent.com/16667688/189328923-de76762c-fe06-4e70-8820-11e9917e7da1.png)

austincunningham commented on controllers/capabilities/application_controller.go:
```diff
@@ -0,0 +1,273 @@
+	err = r.Client().Get(r.Context(), projectMeta, product)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			reqLogger.Info("Product resource not found. Ignoring since object must have been deleted")
```
requires a bit of refactoring. Will take a look now.

austincunningham replied:
moved to applicationReconciler, but needed a seconded api call for account to delete will take a look on monday to see if I can improve this a bit.

austincunningham replied:
need the provideAccount to delete. will have to leave it.

austincunningham commented on controllers/capabilities/application_controller.go:
```diff
@@ -0,0 +1,273 @@
+		return ctrl.Result{}, nil
+	}
+
+	statusReconciler, reconcileErr := r.applicationReconciler(application, product, Account)
```
looks like this has been removed, mark as resolved

eguzki commented:
/retest

austincunningham commented:
/ok-to-test

eguzki commented:
verification steps working :muscle:

eguzki reviewed (changes requested):
Looking good!

I left some comments.

Awesome work testing the app controller logic. 

I see there is some doc for the Application CRD reference. It would be very nice to have some higher level as user guide. Also useful for downstream doc. In https://github.com/3scale/3scale-operator/blob/master/doc/operator-application-capabilities.md, 
* add Application CRD to the index, with link to the ref doc.
* add few examples
* Some "howto" user guide, very similar to the verification steps, so the end user can easily follow steps to deploy an application. 

I realized that we missed this for the ProxyConfigPromote CRD

austincunningham commented on config/samples/capabilities_v1beta1_application.yaml:
```diff
@@ -0,0 +1,11 @@
+  name: application-sample
+spec:
+  accountCRName: "developeraccount-sample"
+  applicationPlanName: "plan01"
```
not outdated just wrong, will amend

austincunningham commented on apis/capabilities/v1beta1/application_types.go:
```diff
@@ -0,0 +1,141 @@
+	// Important: Run "make" to regenerate code after modifying this file
+
+	//AccountCRName name of account custom resource under which the application will be created
+	AccountCRName *corev1.LocalObjectReference `json:"accountCRName"`
```
makes sense, think there was a misunderstanding on how localobjectref was to be populated. Happy with AccountCR and ProduceCR

austincunningham commented on apis/capabilities/v1beta1/application_types.go:
```diff
@@ -0,0 +1,141 @@
+	AccountCRName *corev1.LocalObjectReference `json:"accountCRName"`
+
+	//ProductCRName of product custom resource from which the application plan will be used
+	ProductCRName *corev1.LocalObjectReference `json:"productCRName"`
```
yep will go with ProductCR

eguzki commented:
heads up, after merging https://github.com/3scale/3scale-operator/pull/785, this PR would need to be rebased and some (hopefully few and small) changes should be applied.

austincunningham commented on apis/capabilities/v1beta1/application_types.go:
```diff
@@ -0,0 +1,141 @@
+	Conditions common.Conditions `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type" protobuf:"bytes,2,rep,name=conditions"`
+}
+
+func (b *ApplicationStatus) Equals(other *ApplicationStatus, logger logr.Logger) bool {
```
will amend

austincunningham commented on config/samples/capabilities_v1beta1_application.yaml:
```diff
@@ -0,0 +1,11 @@
+spec:
+  accountCRName: "developeraccount-sample"
+  applicationPlanName: "plan01"
+  productCRName: "product-sample"
```
will amend

austincunningham commented on controllers/capabilities/application_threescale_reconciler.go:
```diff
@@ -0,0 +1,122 @@
+		return nil, err
+	}
+	t.applicationEntity = applicationEntity
+	taskRunner := helper.NewTaskRunner(nil, t.logger)
```
inclined to leave as is it for the timestamps in debug mode

austincunningham commented on controllers/capabilities/application_threescale_reconciler.go:
```diff
@@ -0,0 +1,122 @@
+	}
+
+	// Find application in the list by  name
+	data, err := json.Marshal(t.applicationResource)
```
I think it was done this way for some nil pointer handing when the Status.ID doesn't exist. Will confirm.

austincunningham replied:
yea not needed

austincunningham commented:
/test test-e2e

austincunningham commented:
/test test-e2e

eguzki commented:
Awesome work @austincunningham :medal_military:

eguzki commented:
I spotted small nitpicks, but we can work with them in coming PR's
